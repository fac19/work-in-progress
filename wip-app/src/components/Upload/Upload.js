import React, { Component } from "react";
import Dropzone from "../Dropzone/Dropzone";
import Progress from "../Progress/Progress";
import {
  UploadDiv,
  ContentDiv,
  FilesDiv,
  Actions,
  Title,
  Filename,
  Row,
  CheckIcon,
  ProgressWrapper,
  Button,
} from "./Upload.style.js";

class Upload extends Component {
  constructor(props) {
    super(props);
    this.state = {
      files: [],
      uploading: false,
      uploadProgress: {},
      successfullUploaded: false,
    };

    this.onFilesAdded = this.onFilesAdded.bind(this);
    this.uploadFiles = this.uploadFiles.bind(this);
    this.sendRequest = this.sendRequest.bind(this);
    this.renderActions = this.renderActions.bind(this);
  }

  onFilesAdded(files) {
    this.setState((prevState) => ({
      files: prevState.files.concat(files),
    }));
  }

  async uploadFiles() {
    this.setState({ uploadProgress: {}, uploading: true });
    const promises = [];
    this.state.files.forEach((file) => {
      promises.push(this.sendRequest(file));
    });
    try {
      await Promise.all(promises);

      this.setState({ successfullUploaded: true, uploading: false });
    } catch (e) {
      // Not Production ready! Do some error handling here instead...
      this.setState({ successfullUploaded: true, uploading: false });
    }
  }

  sendRequest(file) {
    return new Promise((resolve, reject) => {
      const req = new XMLHttpRequest();

      req.upload.addEventListener("progress", (event) => {
        if (event.lengthComputable) {
          const copy = { ...this.state.uploadProgress };
          copy[file.name] = {
            state: "pending",
            percentage: (event.loaded / event.total) * 100,
          };
          this.setState({ uploadProgress: copy });
        }
      });

      req.upload.addEventListener("load", (event) => {
        const copy = { ...this.state.uploadProgress };
        copy[file.name] = { state: "done", percentage: 100 };
        this.setState({ uploadProgress: copy });
        resolve(req.response);
      });

      req.upload.addEventListener("error", (event) => {
        const copy = { ...this.state.uploadProgress };
        copy[file.name] = { state: "error", percentage: 0 };
        this.setState({ uploadProgress: copy });
        reject(req.response);
      });

      const formData = new FormData();
      formData.append("file", file, file.name);

      req.open("POST", "http://localhost:8000/upload");
      req.send(formData);
    });
  }

  renderProgress(file) {
    const uploadProgress = this.state.uploadProgress[file.name];
    if (this.state.uploading || this.state.successfullUploaded) {
      return (
        <ProgressWrapper>
          <Progress progress={uploadProgress ? uploadProgress.percentage : 0} />
          <CheckIcon
            className="CheckIcon"
            alt="done"
            src="baseline-check_circle_outline-24px.svg"
            style={{
              opacity:
                uploadProgress && uploadProgress.state === "done" ? 0.5 : 0,
            }}
          />
        </ProgressWrapper>
      );
    }
  }

  renderActions() {
    if (this.state.successfullUploaded) {
      return (
        <Button
          onClick={() =>
            this.setState({ files: [], successfullUploaded: false })
          }
        >
          Clear
        </Button>
      );
    } else {
      return (
        <Button
          disabled={this.state.files.length < 0 || this.state.uploading}
          onClick={this.uploadFiles}
        >
          Upload
        </Button>
      );
    }
  }

  render() {
    return (
      <UploadDiv className="Upload">
        <Title className="Title">Upload Files</Title>
        <ContentDiv className="Content">
          <div>
            <Dropzone
              onFilesAdded={this.onFilesAdded}
              disabled={this.state.uploading || this.state.successfullUploaded}
            />
          </div>
          <FilesDiv className="Files">
            {this.state.files.map((file) => {
              return (
                <Row key={file.name} className="Row">
                  <Filename className="Filename">{file.name}</Filename>
                  {this.renderProgress(file)}
                </Row>
              );
            })}
          </FilesDiv>
        </ContentDiv>
        <Actions className="Actions">{this.renderActions()}</Actions>
      </UploadDiv>
    );
  }
}

export default Upload;
