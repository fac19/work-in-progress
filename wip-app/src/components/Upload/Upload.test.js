import React from "react";
import TestRenderer from "react-test-renderer";
import Upload from "./Upload";

describe("Snapshot test", () => {
  it("renders the correct ui", () => {
    const component = TestRenderer.create(<Upload />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
