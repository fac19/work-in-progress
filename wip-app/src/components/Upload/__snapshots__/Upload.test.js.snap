// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test renders the correct ui 1`] = `
<div
  className="sc-AxgMl jcCOud"
>
  <h3
    className="sc-fzozJi fbndaM"
  >
    Upload Files
  </h3>
  <div
    className="sc-AxheI cDpprM"
  >
    <div>
      <div
        className="sc-AxjAm LyzTN Dropzone "
        onClick={[Function]}
        onDragLeave={[Function]}
        onDragOver={[Function]}
        onDrop={[Function]}
        style={
          Object {
            "cursor": "pointer",
          }
        }
      >
        <input
          className="sc-AxiKw bymOIn FileInput"
          multiple={true}
          onChange={[Function]}
          type="file"
        />
        <img
          alt="upload"
          className="sc-AxirZ dUPqmS Icon"
          src="baseline-cloud_upload-24px.svg"
        />
        <span>
          Upload Files
        </span>
      </div>
    </div>
    <div
      className="sc-Axmtr jNncFm"
    />
  </div>
  <div
    className="sc-AxmLO kpHjMF"
  >
    <button
      className="sc-fznKkj gBWhcO"
      disabled={false}
      onClick={[Function]}
    >
      Upload
    </button>
  </div>
</div>
`;
